Enhance Dynamic Include Path Management in xeus-cpp with XEUS_SEARCH_PATH Support

AhmedFatthy1040 opened the pull request:
# Description

This PR introduces improvements to the `xeus-cpp` library to handle dynamic include paths more effectively. It includes updates to the `interpreter::init_includes` function and the CMakeLists file. These changes aim to enhance flexibility and maintainability by leveraging the `XEUS_SEARCH_PATH` environment variable for managing include directories dynamically.

### **Key Changes**
1. **Updated `interpreter::init_includes` Function:**
   - Parses the `XEUS_SEARCH_PATH` environment variable to include non-standard directories.
   - Adds support for multiple include paths, separated by colons (`:`).

2. **Modified CMakeLists File:**
   - Dynamically builds the `XEUS_SEARCH_PATH` variable by joining target include directories.
   - Includes the `src` directory for non-EMSCRIPTEN builds.
   - Passes the `XEUS_SEARCH_PATH` as a preprocessor definition to the compiler.

These updates simplify the configuration and usage of custom include paths.

Fixes #175 

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [x] New feature 
- [ ] Added/removed dependencies
- [ ] Required documentation updates

AhmedFatthy1040 commented:
I have pushed new changes, please approve the awaiting workflows @vgvassilev

AhmedFatthy1040 commented:
@JohanMabille @anutosh491

anutosh491 commented:
Hi, 

I shall find the time to review this soon !

AhmedFatthy1040 commented:
I have fixed the warnings.

vgvassilev commented on CMakeLists.txt:
```diff
@@ -404,7 +406,21 @@ endif()
 if(XEUS_CPP_BUILD_TESTS)
+    enable_testing()
     add_subdirectory(test)
+    
+    # Find doctest package
+    find_package(doctest REQUIRED)
+    
+    # Test for non-standard include paths
+    add_executable(test_include_paths test/test_include_paths.cpp)
+    target_link_libraries(test_include_paths PRIVATE doctest::doctest)
+    target_include_directories(test_include_paths PRIVATE 
+        ${CMAKE_SOURCE_DIR}/test/custom_includes
+        ${DOCTEST_INCLUDE_DIR}
+    )
+    target_compile_definitions(test_include_paths PRIVATE DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN)
+    add_test(NAME test_include_paths COMMAND test_include_paths)
```
This block should go in the test folder.

vgvassilev commented on src/xinterpreter.cpp:
```diff
@@ -357,7 +361,28 @@ __get_cxx_version ()
+        // Add the standard include path
         Cpp::AddIncludePath((xeus::prefix_path() + "/include/").c_str());
+
+        // Get include paths from environment variable
+        const char* non_standard_paths = std::getenv("XEUS_SEARCH_PATH");
+        if (!non_standard_paths) {
+            non_standard_paths = "";
+        }
+
+        if (std::strlen(non_standard_paths) > 0)
+        {
+            // Split the paths by colon ':' and add each one
+            std::istringstream stream(non_standard_paths);
+            std::string path;
+            while (std::getline(stream, path, ':'))
+            {
+                if (!path.empty())
+                {
+                    Cpp::AddIncludePath(path.c_str());
+                }
+            }
+        }
```
We can simplify this by a lot, reduce the mount braces and also consider non-unix path delimiters.

AhmedFatthy1040 replied:
Done

AhmedFatthy1040 commented:
I have fixed the clang-format problem.

vgvassilev commented on src/xinterpreter.cpp:
```diff
@@ -357,7 +362,32 @@ __get_cxx_version ()
+        // Split and add each non-empty path
+        std::istringstream stream(non_standard_paths);
+        std::string path;
+        while (std::getline(stream, path, path_separator))
+        {
+            if (!path.empty())
+            {
+                Cpp::AddIncludePath(path.c_str());
+            }
+        }
```
```suggestion
        while (std::getline(stream, path, path_separator))
            if (!path.empty())
                Cpp::AddIncludePath(path.c_str());
```

AhmedFatthy1040 replied:
Apologies, I initially thought the braces were necessary for the clang-tidy format check to pass. I've now edited and correctly formatted the file.

AhmedFatthy1040 commented:
IDK why it's failing now.
![White Cat](https://github.com/user-attachments/assets/60696e35-3e61-40b3-83ac-c4994f13d79b)

mcbarton commented:
@AhmedFatthy1040 please squash the commits of the PR and perform git-clang-format on the single commit. Then the ci should pass. Then another review can be made.

AhmedFatthy1040 commented:
@mcbarton  I previously addressed this issue by adding braces here. However, @vgvassilev recommended removing them for code simplification, which I implemented. Unfortunately, this removal triggered a failure in the git clang-format check. I have now corrected the issue by re-adding the braces. Please approve the pending workflows.
![Screenshot from 2024-12-19 20-27-06](https://github.com/user-attachments/assets/8016663a-b38f-47a3-88f0-cc7934cec1a6)

mcbarton commented:
> @mcbarton I previously addressed this issue by adding braces here. However, @vgvassilev recommended removing them for code simplification, which I implemented. Unfortunately, this removal triggered a failure in the git clang-format check. I have now corrected the issue by re-adding the braces. Please approve the pending workflows. ![Screenshot from 2024-12-19 20-27-06](https://private-user-images.githubusercontent.com/91102592/397492111-8016663a-b38f-47a3-88f0-cc7934cec1a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ2MzQxNTUsIm5iZiI6MTczNDYzMzg1NSwicGF0aCI6Ii85MTEwMjU5Mi8zOTc0OTIxMTEtODAxNjY2M2EtYjM4Zi00N2EzLTg4ZjAtY2M3OTM0Y2VjMWE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE5VDE4NDQxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1MTcwYjZlYjVlMmVlOGU2OTE1MDdiZGE0MjkxYTAyYjI5NjYzN2JmMDE3ZGNhOTY3YmFhYWJkNjI5MmU5NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pPYIOnbKy7VF5qq7nKSeUWBvAytd5QIc5EqF-SLqI9k)

@AhmedFatthy1040 Please revert this change. Squash the commits into 1 commit, then perform git-clang-format on this single commit. Then well see if the workflows pass. We can then review it from that point.

AhmedFatthy1040 commented:
Sorry for the mistake in the last commit, i am fixing it now.

AhmedFatthy1040 commented:
> > @mcbarton I previously addressed this issue by adding braces here. However, @vgvassilev recommended removing them for code simplification, which I implemented. Unfortunately, this removal triggered a failure in the git clang-format check. I have now corrected the issue by re-adding the braces. Please approve the pending workflows. ![Screenshot from 2024-12-19 20-27-06](https://private-user-images.githubusercontent.com/91102592/397492111-8016663a-b38f-47a3-88f0-cc7934cec1a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ2MzQxNTUsIm5iZiI6MTczNDYzMzg1NSwicGF0aCI6Ii85MTEwMjU5Mi8zOTc0OTIxMTEtODAxNjY2M2EtYjM4Zi00N2EzLTg4ZjAtY2M3OTM0Y2VjMWE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE5VDE4NDQxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1MTcwYjZlYjVlMmVlOGU2OTE1MDdiZGE0MjkxYTAyYjI5NjYzN2JmMDE3ZGNhOTY3YmFhYWJkNjI5MmU5NGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pPYIOnbKy7VF5qq7nKSeUWBvAytd5QIc5EqF-SLqI9k)
> 
> @AhmedFatthy1040 Please revert this change. Squash the commits into 1 commit, then perform git-clang-format on this single commit. Then well see if the workflows pass. We can then review it from that point.

okay

AhmedFatthy1040 commented:
I found out that i was using an old version of clang-format, so i updated to version 18 as it is the version used in the CI/CD pipeline, and formatted the code again @mcbarton

AhmedFatthy1040 commented:
I will try to fix the clang-tidy suggestions tomorrow. Additionally, there was a block of code that was refactored, but it got deleted by mistake when I reset some commits. I will fix that as well.
